enhancements/workload-partitioning/cpu-limits-workload-partitioning.md

```diff
@@ -0,0 +1,417 @@
+`CPUQuota` option.
+
+[workloads.go](https://github.com/cri-o/cri-o/blob/f243ba712d58d106dea1ba7adf33ed0911a3e563/pkg/config/workloads.go#L45-L50)
+
+```go
+type Resources struct {
+	// Specifies the number of CPU shares this pod has access to.
+	CPUShares uint64 `json:"cpushares,omitempty"`
+    // Specifies the CPU quota this pod is limited to in microseconds.
+	CPUQuota int64 `json:"cpuquota,omitempty"`
+	// Specifies the cpuset this pod has access to.
+	CPUSet string `json:"cpuset,omitempty"`
+}
+```
+
+We'll need to make sure the `MutateSpec` is updated to modify the CPU Resources
+with the correct quota.
+
+[workloads.go](https://github.com/cri-o/cri-o/blob/f243ba712d58d106dea1ba7adf33ed0911a3e563/pkg/config/workloads.go#L173-L183)
+
+```go
+func (r *Resources) MutateSpec(specgen *generate.Generator) {
+	if r == nil {
+		return
+	}
+	if r.CPUSet != "" {
+		specgen.SetLinuxResourcesCPUCpus(r.CPUSet)
+	}
+	if r.CPUShares != 0 {
+		specgen.SetLinuxResourcesCPUShares(r.CPUShares)
+	}
+    if r.CPUQuota != 0 {
+		specgen.SetLinuxResourcesCPUQuota(r.CPUQuota)
+	}
+}
+```
+
+The Cgroup manager will then be updated when calling `setWorkloadsSettings` to
+utilize the top level config when creating the Cgroup.
+
+[cgroupfs_linux.go](https://github.com/cri-o/cri-o/blob/f243ba712d58d106dea1ba7adf33ed0911a3e563/internal/config/cgmgr/cgroupfs_linux.go#L211-L234)
+
+```go
+func setWorkloadSettings(cgPath string, resources *rspec.LinuxResources) (err error) {
+	if resources.CPU == nil {
+		return nil
+	}
+	cg := &cgcfgs.Cgroup{
+		Path: "/" + cgPath,
+		Resources: &cgcfgs.Resources{
+			SkipDevices: true,
+			CpusetCpus:  resources.CPU.Cpus,
+		},
+		Rootless: rootless.IsRootless(),
+	}
+	if resources.CPU.Shares != nil {
+		cg.Resources.CpuShares = *resources.CPU.Shares
+	}
+	if resources.CPU.Quota != nil {
+		cg.Resources.CpuQuota = *resources.CPU.Quota
+	}
+
+	mgr, err := libctrCgMgr.New(cg)
+	if err != nil {
+		return err
+	}
+	return mgr.Set(cg.Resources)
+}
+```
+
+This will all be exposed to the user via the `toml` configuration.
+
+```toml
+[crio.runtime.workloads.management]
+activation_annotation = "target.workload.openshift.io/management"
+annotation_prefix = "resources.workload.openshift.io"
+resources = { "cpushares" = 0,  "cpuquota" = 0, "cpuset" = "0-1,52-53" }
+```
+
+### Admission Webhook
+
+The admission webhook will be updated to use the CPU limit resource information
+to add the annotations to the Pod for each container. Pods with cpu limits will
+no longer be ignored and their limits will correctly be represented with the
+`management.workload.openshift.io/cores` in the same way that `requests` are.
+Containers that do not include `limits` will not have a `cpuquota` set. We will
+be utilizing the helper function in
+[cm.MilliCPUToQuota](https://github.com/openshift/kubernetes/blob/0e0d15b865ffc36177dc8770b4723dc14476a630/pkg/kubelet/cm/helpers_linux.go#L58-L82)
```

Comment:
It is unlikely but possible that the  cpu.cfs_period_us is overridden, We should pass down the specified limit and have crio compute the value based on the period specified on the node.

Reply:
I see, I went a head and altered the approach a bit here to make use of this. Thanks @browsell !